APE26.rst

```diff
@@ -0,0 +1,325 @@
+and attributes are defined on ``SkyCoord`` (and ``Coordinate``)
+proper, so sphinx will know how to typeset those, while type
+checkers can help users in finding and using them properly. It
+will also be easier: following the Zen of Python, there should be
+one clear way to do something. The present overlap leads to
+confusion wherein beginner users end up creating
+``BaseCoordinateFrame`` objects such as ``ICRS``, when the docs
+are clear that these are for more advanced users and that
+``SkyCoord`` is to be preferred. The system will also be less
+fragile. For example, if users manipulate the internal workings
+of ``SkyCoord`` (which is discouraged but possible), the
+coordinate data can become decoupled from the caching that
+``SkyCoord`` performs for speed. With these proposed changes
+users will have a more clear, introspectable, and robust system.
+
+Finished Product
+----------------
+The end result of the implementation of this APE are be two separate
+hierarchies of classes: reference frame classes and coordinate
+classes which bring together a reference frame and coordinate
+data. We discuss each class type in turn.
+
+Coordinate frame classes only hold information pertaining to
+the reference frame they represent and never actual coordinate
+data in that reference frame. This is consistent with our
+mathematical framework, the reference frame mediates how
+coordinate data is understood (e.g. distance measures) or
+interacts (e.g. separation from other coordinates), but the
+coordinate data itself is actually independent of that
+information.
+
+Classes like ``SkyCoord`` will be composed structures bringing
+together the reference frame (an instance of a ``BaseFrame``
+subclass) and the coordinate data (``BaseRepresentation``
+objects). We also introduce a new class, ``Coordinate``, which
+is akin to ``SkyCoord`` (containing both frame and data), but
+without extra features like keeping frame attributes not
+associated with the current frame, caching and flexible input
+parsing. In this way ``Coordinate`` operates very similarly to
+the current ``BaseCoordinateFrame`` objects when they have
+data, and is meant to be their direct replacement in the new
+framework as well as a more lightweight and performant
+alternative to ``SkyCoord``.
+
+We illustrate the new framework with the following pseudocode.
+
+.. code-block:: python
+
+    class BaseFrame:
+        ...
+
+    class ICRSFrame(BaseFrame):
+        pass  # no frame attributes
+
+    class FK5Frame(BaseFrame):
+        equinox: TimeAttribute
+
+    # ------
+
+    class BaseCoordinate:
+        frame: BaseFrame
+        data: BaseRepresentation
+
+    class Coordinate(BaseCoordinate):
+        ...  # it's fast.
+
+    class SkyCoord(BaseCoordinate):
+
+        def __init__(...):  # flexible input parsing
+            ...
+
+Branches and pull requests
+--------------------------
+No direct progress on these changes has yet occurred. Discussion
```

Comment:
How much effort would it be to do a prototype implementation? Are the proposed changes here actually going to amount to a lot of lines of code if we exclude changes to tests, docs, and deprecation warnings/layers?

Reply:
The prototype has been implemented as is given in a subsequent comment. The proposed changes here would remove many more lines of legacy code than new lines added (the new `Coordinate` class for example is by design minimal), and the bulk of the refactoring is in moving data-aware methods/attributes/properties out of `SkyCoord` and into a new base class (`BaseCoordinate`).